components/script/dom/bluetoothremotegattserver.rs

```diff
@@ -124,7 +133,54 @@ impl BluetoothRemoteGATTServerMethods for BluetoothRemoteGATTServer {
                 println!("{}", error);
                 None
             },
-            _ => unreachable!()
+            _ => unreachable!(),
+        }
+    }
+
+    // https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetoothremotegattserver-getprimaryservices
+    fn GetPrimaryServices(&self, service: Option<StringOrUnsignedLong>)
+                          -> Option<Vec<Root<BluetoothRemoteGATTService>>> {
+        let uuid: Option<String> = match service {
+            Some(s) => match BluetoothUUID::GetService(self.global().r(), s.clone()) {
+                Ok(domstring) => Some(domstring.to_string()),
+                Err(_) => None,
+            },
+            None => None,
+        };
+        let mut services: Vec<Root<BluetoothRemoteGATTService>> = vec!();
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.get_bluetooth_thread().send(
+            BluetoothMethodMsg::GetPrimaryServices(String::from(self.Device().Id()), uuid, sender)).unwrap();
+        let services_vec = receiver.recv().unwrap();
+        match services_vec {
+            BluetoothObjectMsg::BluetoothServices {
+                services_vec
+            } => {
+                for s in services_vec {
+                    match s {
+                        BluetoothObjectMsg::BluetoothService {
+                            uuid,
+                            is_primary,
+                            instance_id,
+                        } => {
+                            services.push(BluetoothRemoteGATTService::new(self.global().r(),
+                                                                          &self.device.get(),
+                                                                          DOMString::from(uuid),
+                                                                          is_primary,
+                                                                          instance_id))
+                        },
+                        _ => unreachable!(),
+                    }
+                }
+                Some(services)
+            },
+            BluetoothObjectMsg::Error {
+                error
+            } => {
+                println!("{}", error);
```

Comment:
Our next patch will be an error handling related, so you can ignore this for now.